Microsoft.Azure.CosmosRepository/src/ChangeFeed/DefaultConatinerChangeFeedProcessor.cs

```diff
@@ -0,0 +1,148 @@
+
+            _processor = itemContainer.GetChangeFeedProcessorBuilder<JObject>("cosmos-repository-pattern-processor",
+                    (changes, token) => OnChanges(changes, token, itemContainer.Id))
+                .WithLeaseContainer(leaseContainer)
+                .WithInstanceName(_changeFeedOptions.InstanceName)
+                .WithErrorNotification((token, exception) => OnError(exception, itemContainer.Id))
+                .Build();
+
+            _logger.LogInformation("Starting change feed processor for container {ContainerName}", itemContainer.Id);
+
+            await _processor.StartAsync();
+
+            _logger.LogInformation("Successfully started change feed processor for container {ContainerName}",
+                itemContainer.Id);
+        }
+
+        internal async Task OnChanges(IReadOnlyCollection<JObject> changes, CancellationToken cancellationToken,
+            string containerName)
+        {
+            _logger.LogDebug("Detected changes for container {ContainerName} total ({ChangesCount})",
+                containerName, changes.Count);
+
+            foreach (JObject change in changes)
+            {
+                if (!change.TryGetValue("type", out JToken type))
+                {
+                    //TODO: _logger.LogWarning(...) here? Should this be turn on/off thing? A user might not have created documents with the type field ?!
+                    continue;
+                }
+
+                Type itemType = ItemTypes.FirstOrDefault(x => x.Name == type.Value<string>());
+
+                if (itemType is null)
+                {
+                    _logger.LogDebug(
+                        "No change feed processor registered for type {ItemType} in container {ContainerName}",
+                        type.Value<string>(), containerName);
+
+                    continue;
+                }
+
+                await InvokeHandler(itemType, change, cancellationToken);
+            }
+        }
+
+        private async Task InvokeHandler(Type itemType, JObject instance, CancellationToken cancellationToken)
+        {
+            object item = instance.ToObject(itemType);
+
+            Type handlerType = null;
+
+            if (Handlers.ContainsKey(itemType) is false)
+            {
+                handlerType = typeof(IItemChangeFeedProcessor<>).MakeGenericType(itemType);
+                Handlers[itemType] = handlerType;
+            }
+
+            handlerType ??= Handlers[itemType];
+
+            IEnumerable<object> handlers = _serviceProvider.GetServices(handlerType).ToList();
+
+            _logger.LogDebug("Invoking IItemChangeFeedProcessor's ({ProcessorsCount}) for item type {ItemType}",
+                handlers.Count(), itemType);
+
+            foreach (object handler in handlers)
+            {
+                try
+                {
+                    object response = handlerType.GetMethod("HandleAsync")?
```

Comment:
```suggestion
                    object response = handlerType.GetMethod("HandleAsync")?
```

I'm not super excited by the magic string for the method name here. If it changes, this will break - it's a bit fragile. Since we're down casting to an `object` collection from the service provider of a given type, does that raise any alarms from a DI perspective? Trying to think through this a bit more. What other implications are there? When you step-debug through this, does it do exactly what you imagine?

Reply:
Yes I was not over the moon about this either, but if I am honest I think it's a trade-off we may have to take, there was no way (at least that I could see) when having one change feed processor per container, that could handle multiple `IItem`s to flow down a generic `TItem` property to neatly resolve this from DI by calling something like `_provider.GetRequiredService<IItemChangeFeedProcessor<TItem>>();` which was my initial intention, this was the best I could do, as far as debugging it behaves the same way you would see the normal call as shown but I agree the floating string feels a bit meh, however, we control the interface so I think that makes it a lot safer. Thoughts?

Reply:
That's fair, I was just pushing on it a bit as I was curious what alternatives have been looked at - and how you felt about it. Thanks

Reply:
Okay cool, I agree with you tho, doesn't look super clean but not sure it does much harm by the same token.